src/Init/System/IOError.lean

```diff
@@ -0,0 +1,193 @@
+def mkPermissionDeniedFile : String → String → IO.Error :=
+permissionDenied ∘ some
+
+@[export lean_mk_io_error_resource_exhausted_file]
+def mkResourceExhaustedFile : String → String → IO.Error :=
+resourceExhausted ∘ some
+
+@[export lean_mk_io_error_unsupported_operation]
+def mkUnsupportedOperation : String → IO.Error :=
+unsupportedOperation
+
+@[export lean_mk_io_error_resource_exhausted]
+def mkResourceExhausted : String → IO.Error :=
+resourceExhausted none
+
+@[export lean_mk_io_error_already_exists]
+def mkAlreadyExists : String → IO.Error :=
+alreadyExists
+
+@[export lean_mk_io_error_inappropriate_type]
+def mkInappropriateType : String → IO.Error :=
+inappropriateType none
+
+@[export lean_mk_io_error_no_such_thing]
+def mkNoSuchThing : String → IO.Error :=
+noSuchThing none
+
+@[export lean_mk_io_error_resource_vanished]
+def mkResourceVanished : String → IO.Error :=
+resourceVanished
+
+@[export lean_mk_io_error_resource_busy]
+def mkResourceBusy : String → IO.Error :=
+resourceBusy
+
+@[export lean_mk_io_error_invalid_argument]
+def mkInvalidArgument : String → IO.Error :=
+invalidArgument none
+
+@[export lean_mk_io_error_other_error]
+def mkOtherError : String → IO.Error :=
+otherError
+
+@[export lean_mk_io_error_permission_denied]
+def mkPermissionDenied : String → IO.Error :=
+permissionDenied none
+
+@[export lean_mk_io_error_hardware_fault]
+def mkHardwareFault : String → IO.Error :=
+hardwareFault
+
+@[export lean_mk_io_error_unsatisfied_constraints]
+def mkUnsatisfiedConstraints : String → IO.Error :=
+unsatisfiedConstraints
+
+@[export lean_mk_io_error_illegal_operation]
+def mkIllegalOperation : String → IO.Error :=
+illegalOperation
+
+@[export lean_mk_io_error_protocol_error]
+def mkProtocolError : String → IO.Error :=
+protocolError
+
+@[export lean_mk_io_error_time_expired]
+def mkTimeExpired : String → IO.Error :=
+timeExpired
+
+private def downCaseFirst (s : String) : String := s.set 0 (s.get 0).toLower
+
+def fopenErrorToString (gist fn : String) : Option String → String
+| some details => gist ++ ": " ++ downCaseFirst details ++ "\n  file: " ++ fn
+| none => gist ++ "\n  file: " ++ fn
+
+def otherErrorToString (gist : String) : Option String → String
+| some details => gist ++ ": " ++ downCaseFirst details
+| none => gist
+
+@[export lean_io_error_to_string]
+def toString : IO.Error → String
+| unexpectedEof                       => "End of file"
+| inappropriateType (some fn) details => fopenErrorToString "Inappropriate type" fn details
+| inappropriateType none details      => otherErrorToString "Inappropriate type" details
+| interrupted fn details              => fopenErrorToString "Interrupted system call" fn details
+| invalidArgument (some fn) details   => fopenErrorToString "Invalid argument" fn details
+| invalidArgument none details        => otherErrorToString "Invalid argument" details
+| noFileOrDirectory fn details        => fopenErrorToString "No such file or directory" fn none
```

Comment:
perhaps replace `details` with `_` to make it more noticeable that it gets ignored